src/parser.rs

```diff
@@ -106,59 +45,583 @@ impl<'a> Message<'a> {
             }) => Ok(Message::MagicCookieServer),
 
             Ok(header) => {
-                let payload_len = header.length as usize - 8; // PRS_SOMEIP_00042
+                let payload_len = header.payload_len();
                 if input.len() < HEADER_LENGTH + payload_len {
-                    Err(Error::NotEnoughData {
+                    return Err(Error::NotEnoughData {
                         min: HEADER_LENGTH + payload_len,
                         actual: input.len(),
-                    })
+                    });
+                }
+
+                let payload = &input[HEADER_LENGTH..HEADER_LENGTH + payload_len];
+
+                if header.is_sd() {
+                    let mut buffer = Cursor::new(payload);
+                    Ok(Message::SdMessage(
+                        header,
+                        SdPayload::from_reader(&mut buffer)?,
+                    ))
                 } else {
-                    let payload = &input[HEADER_LENGTH..HEADER_LENGTH + payload_len];
-                    Ok(Message::Message(header, payload))
+                    Ok(Message::RpcMessage(header, payload))
                 }
             }
             Err(e) => Err(e),
         }
     }
 }
 
+impl Header {
+    /// Parse a header from a byte slice.
+    pub fn from_slice(input: &[u8]) -> Result<Header, Error> {
+        if input.len() < HEADER_LENGTH {
+            return Err(Error::NotEnoughData {
+                min: HEADER_LENGTH,
+                actual: input.len(),
+            });
+        }
+
+        let mut buffer = Cursor::new(input);
+        Header::from_reader(&mut buffer)
+    }
+
+    fn from_reader<R: Read>(reader: &mut R) -> Result<Header, Error> {
```

Comment:
donno.. if it should be maybe consistent on Message, Payloads etc too. Like parser provide "from_bytes/from_reader" and serializer "to_bytes/to_writer" for all?!

Reply:
I think in order to work efficiently the use of `Message` is omitted. You know in when you expect a RPC message or a SD message. You never receive both types on the same channel/socket/whatever.
I'd like:

```
Header::from_reader/slice
SdPayload::from_reader/slice
```

The SD entries are IMHO not needed because you probably never need them standalone serialized/deserialized.

Reply:
"You know in when you expect a RPC message or a SD message.".. not true if you parse messages from eg a pcap trace..

Reply:
OK. So - after you have the header. 

I'm fine with the `Messages` enum - but regard it at totally useless :-)